recorder.go

```diff
@@ -313,29 +315,71 @@ func translateDurationFromOldesYoungest(ot time.Time, yt time.Time) uint64 {
 func translateTags(tags ot.Tags) []*cpb.KeyValue {
 	kvs := make([]*cpb.KeyValue, 0, len(tags))
 	for key, tag := range tags {
-		kv := cpb.KeyValue{Key: key}
-		switch v := tag.(type) {
-		case string:
-			kv.Value = &cpb.KeyValue_StringValue{v}
-		case int:
-			kv.Value = &cpb.KeyValue_IntValue{int64(v)}
-		case float64:
-			kv.Value = &cpb.KeyValue_DoubleValue{v}
-		case bool:
-			kv.Value = &cpb.KeyValue_BoolValue{v}
-		default:
-			glog.Infof("value: %v, %T, is an unsupported type, and has been converted to string", v, v)
-			// TODO: use reflection so that not all custom types have to be converted to string
-			kv.Value = &cpb.KeyValue_StringValue{fmt.Sprint(v)}
+		kv := convertToKeyValue(key, tag)
+		kvs = append(kvs, kv)
+	}
+	return kvs
+}
+
+func convertToKeyValue(k string, value interface{}) *cpb.KeyValue {
+	kv := cpb.KeyValue{Key: k}
+	switch v := value.(type) {
+	case string:
+		kv.Value = &cpb.KeyValue_StringValue{v}
+	case int:
+		kv.Value = &cpb.KeyValue_IntValue{int64(v)}
+	case float64:
+		kv.Value = &cpb.KeyValue_DoubleValue{v}
+	case bool:
+		kv.Value = &cpb.KeyValue_BoolValue{v}
+	default:
+		glog.Infof("value: %v, %T, is an unsupported type, and has been converted to string", v, v)
+		glog.Infof(" the key is %v ", k)
+		// TODO: use reflection so that not all custom types have to be converted to string
+		kv.Value = &cpb.KeyValue_StringValue{fmt.Sprint(v)}
+	}
+	return &kv
+}
+
+func translatePayload(pl interface{}, kvs []*cpb.KeyValue) []*cpb.KeyValue {
+	switch p := pl.(type) {
+	case map[string]interface{}:
+		for k, v := range p {
+			kvs = append(kvs, convertToKeyValue(k, v))
+			glog.Infof("the map is looks like {%v : %v}", k, v)
 		}
-		kvs = append(kvs, &kv)
+	case []interface{}:
+		for _, v := range p {
+			kvs = append(kvs, convertToKeyValue(payloadKey, v))
+		}
+	case nil:
+		return nil
+	default:
+		kvs = append(kvs, convertToKeyValue(payloadKey, p))
 	}
 	return kvs
 }
 
-// TODO: Implement once OT logs have been updated
-func translateLogs(log []ot.LogData) []*cpb.Log {
-	return nil
+func translateEventAndPayload(e string, pl interface{}) []*cpb.KeyValue {
+	var kvs []*cpb.KeyValue
+	kvs = append(kvs, &cpb.KeyValue{Key: messageKey, Value: &cpb.KeyValue_StringValue{e}})
+	return translatePayload(pl, kvs)
```

Comment:
The above three lines could be written 
return translatePayload(pl, []*cpb.KeyValue{&cpb.KeyValue{Key: messageKey, Value: &cpb.KeyValue_StringValue{e}}})

(now that I read it, not sure it's better)